### ISSUE TITLE
Code: 44. DB::Exception: Nullable(String) cannot be inside Nullable column: while executing

### ISSUE THREAD

**Issue by ghost** [ClickHouse/ClickHouse#46128]
**Describe the issue**
I get the following error "Code: 44. DB::Exception: Nullable(String) cannot be inside Nullable column: while executing" when using a replace inside an if statement when the column inside the replace is nullable. Repro below.

**How to reproduce**
* We upgraded from 22.2.3.5 to 22.8.11.15
* Which ClickHouse server versions are incompatible

Settings

Repro:
```sql
-- auto-generated definition
create table `example`
(
    Book LowCardinality(String),
    `Run Date`                Date32,
    M LowCardinality(Nullable(String)),
    Q LowCardinality(Nullable(String))
)
    engine = MergeTree PARTITION BY toYYYYMMDD(`Run Date`)
        ORDER BY (`Run Date`, Book);

INSERT INTO example VALUES ('1', toDate32('20230101'), '1', '2');
INSERT INTO example VALUES ('2', toDate32('20230101'), '2', '3');

SELECT multiIf(M = '1',  replace(Q, '1', 'A'), Q) FROM example; -- doesn't work
SELECT if(M = '1',  replace(Q, '1', 'A'), Q) FROM example; -- doesn't work

SELECT multiIf(M = '1',  coalesce(replace(Q, '1', 'A'), ''), Q) FROM example; -- works

SELECT multiIf(true,  replace(Q, '1', 'A'), Q) FROM example; -- works also

SELECT multiIf(M = '1',  replace(Q, '1', 'A'), 'works') FROM example; -- this works too??

SELECT multiIf(M = '1',  replace(Q, '1', 'A'), coalesce(Q, '')) FROM example; -- this works too??

create table `example2`
(
    Book LowCardinality(String),
    `Run Date`                Date32,
    M LowCardinality(String),
    Q LowCardinality(String)
)
    engine = MergeTree PARTITION BY toYYYYMMDD(`Run Date`)
        ORDER BY (`Run Date`, Book);

INSERT INTO example2 VALUES ('1', toDate32('20230101'), '1', '2');
INSERT INTO example2 VALUES ('2', toDate32('20230101'), '2', '3');

SELECT multiIf(M = '1',  replace(Q, '1', 'A'), Q) FROM example2; -- works when columns not nullable

```

**Error message and/or stacktrace**
Code: 44. DB::Exception: Nullable(String) cannot be inside Nullable column: while executing 'FUNCTION if(equals(M, '1') :: 4, replaceAll(Q, '1', 'A') :: 0, Q :: 1) -> if(equals(M, '1'), replaceAll(Q, '1', 'A'), Q) Nullable(String) : 3'. (ILLEGAL_COLUMN) (version 22.8.11.15 (official build))
 
Stacktrace from database server:
2023.02.07 17:32:05.474370 [ 20258 ] {0a753bd7-ee5d-4d99-b994-428403f0013e} <Error> executeQuery: Code: 44. DB::Exception: Nullable(String) cannot be inside Nullable column: while executing 'FUNCTION if(equals(M, '1') :: 4, replaceAll(Q, '1', 'A') :: 0, Q :: 1) -> if(equals(M, '1'), replaceAll(Q, '1', 'A'), Q) Nullable(String) : 3'. (ILLEGAL_COLUMN) (version 22.8.11.15 (official build)) (from 10.246.228.30:56996) (in query: SELECT multiIf(M = '1', replace(Q, '1', 'A'), Q) FROM example FORMAT TabSeparatedWithNamesAndTypes), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xa3f9b5a in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
1. DB::ColumnNullable::ColumnNullable(COW<DB::IColumn>::mutable_ptr<DB::IColumn>&&, COW<DB::IColumn>::mutable_ptr<DB::IColumn>&&) @ 0x1529f374 in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
2. ? @ 0xf1b87b0 in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
3. ? @ 0xf1b5cd6 in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
4. DB::FunctionToExecutableFunctionAdaptor::executeImpl(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long) const @ 0xcbaa3ce in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
5. DB::IExecutableFunction::executeWithoutLowCardinalityColumns(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x13ceb847 in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
6. DB::IExecutableFunction::executeWithoutSparseColumns(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x13cec13d in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
7. DB::IExecutableFunction::execute(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&, std::__1::shared_ptr<DB::IDataType const> const&, unsigned long, bool) const @ 0x13ced5e9 in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
8. DB::ExpressionActions::execute(DB::Block&, unsigned long&, bool) const @ 0x1498e864 in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
9. DB::ExpressionTransform::transform(DB::Chunk&) @ 0x15ed001c in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
10. DB::ISimpleTransform::transform(DB::Chunk&, DB::Chunk&) @ 0x11c926b0 in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
11. DB::ISimpleTransform::work() @ 0x15cd9529 in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
12. DB::ExecutionThreadContext::executeTask() @ 0x15cf94aa in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
13. DB::PipelineExecutor::executeStepImpl(unsigned long, std::__1::atomic<bool>*) @ 0x15ceda1b in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
14. ? @ 0x15cf02b9 in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
15. ThreadPoolImpl<std::__1::thread>::worker(std::__1::__list_iterator<std::__1::thread, void*>) @ 0xa4bb6a8 in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
16. ? @ 0xa4be8bd in /local/data/scratch/rl-eng-dev/conda-data/envs/clickhouse-8cb7a43f82c02e0d6766d561a119d8f9/clickhouse/usr/bin/clickhouse
17. start_thread @ 0x7ea5 in /usr/lib64/libpthread-2.17.so
18. __clone @ 0xfeb0d in /usr/lib64/libc-2.17.so

**Additional context**

**Comment by tezhm** [comment]
Getting something kind of similar on version 22.8.15.23 (have a low cardinality column with nullable string):

```
2023.03.24 05:37:34.824861 [ 48 ] {fe140540-e718-4e19-9382-ecd94badbcde} <Error> DynamicQueryHandler: Code: 43. DB::Exception: Nested type LowCardinality(String) cannot be inside Nullable type. (ILLEGAL_TYPE_OF_ARGUMENT), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, int, bool) @ 0xa3ffe9a in /usr/bin/clickhouse
1. DB::DataTypeNullable::DataTypeNullable(std::__1::shared_ptr<DB::IDataType const> const&) @ 0x141f71bf in /usr/bin/clickhouse
2. ? @ 0xcb33832 in /usr/bin/clickhouse
3. DB::CastOverloadResolverImpl<(DB::CastType)2, true, DB::CastInternalOverloadName, DB::CastInternalName>::getReturnTypeImpl(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&) const @ 0xce6e714 in /usr/bin/clickhouse
4. DB::IFunctionOverloadResolver::getReturnTypeWithoutLowCardinality(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&) const @ 0x13d03d93 in /usr/bin/clickhouse
5. DB::IFunctionOverloadResolver::getReturnType(std::__1::vector<DB::ColumnWithTypeAndName, std::__1::allocator<DB::ColumnWithTypeAndName> > const&) const @ 0x13d03999 in /usr/bin/clickhouse
6. DB::castColumnAccurateOrNull(DB::ColumnWithTypeAndName const&, std::__1::shared_ptr<DB::IDataType const> const&) @ 0x1518dd91 in /usr/bin/clickhouse
7. DB::KeyCondition::transformConstantWithValidFunctions(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, unsigned long&, std::__1::shared_ptr<DB::IDataType const>&, DB::Field&, std::__1::shared_ptr<DB::IDataType const>&, std::__1::function<bool (DB::IFunctionBase&, DB::IDataType const&)>) const @ 0x1583afff in /usr/bin/clickhouse
8. DB::KeyCondition::canConstantBeWrappedByFunctions(DB::KeyCondition::Tree const&, unsigned long&, std::__1::shared_ptr<DB::IDataType const>&, DB::Field&, std::__1::shared_ptr<DB::IDataType const>&) @ 0x1583c98a in /usr/bin/clickhouse
9. DB::KeyCondition::tryParseAtomFromAST(DB::KeyCondition::Tree const&, std::__1::shared_ptr<DB::Context const>, DB::Block&, DB::KeyCondition::RPNElement&) @ 0x15839776 in /usr/bin/clickhouse
10. DB::KeyCondition::traverseAST(DB::KeyCondition::Tree const&, std::__1::shared_ptr<DB::Context const>, DB::Block&) @ 0x15837126 in /usr/bin/clickhouse
11. DB::KeyCondition::traverseAST(DB::KeyCondition::Tree const&, std::__1::shared_ptr<DB::Context const>, DB::Block&) @ 0x15837089 in /usr/bin/clickhouse
12. DB::KeyCondition::traverseAST(DB::KeyCondition::Tree const&, std::__1::shared_ptr<DB::Context const>, DB::Block&) @ 0x15837089 in /usr/bin/clickhouse
13. DB::KeyCondition::KeyCondition(std::__1::shared_ptr<DB::IAST> const&, std::__1::vector<std::__1::shared_ptr<DB::IAST>, std::__1::allocator<std::__1::shared_ptr<DB::IAST> > > const&, std::__1::shared_ptr<DB::TreeRewriterResult const>, std::__1::shared_ptr<DB::PreparedSets>, std::__1::shared_ptr<DB::Context const>, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, std::__1::shared_ptr<DB::ExpressionActions> const&, bool, bool) @ 0x15836a98 in /usr/bin/clickhouse
14. DB::MergeTreeDataSelectExecutor::filterPartsByPartition(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const> > >&, std::__1::optional<std::__1::unordered_set<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > > const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::MergeTreeData const&, DB::SelectQueryInfo const&, std::__1::shared_ptr<DB::Context const> const&, std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, long, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, long> > > const*, Poco::Logger*, std::__1::vector<DB::ReadFromMergeTree::IndexStat, std::__1::allocator<DB::ReadFromMergeTree::IndexStat> >&) @ 0x15949498 in /usr/bin/clickhouse
15. DB::ReadFromMergeTree::selectRangesToRead(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const> > >, std::__1::shared_ptr<DB::PrewhereInfo> const&, DB::ActionDAGNodes const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, std::__1::shared_ptr<DB::StorageInMemoryMetadata const> const&, DB::SelectQueryInfo const&, std::__1::shared_ptr<DB::Context const>, unsigned int, std::__1::shared_ptr<std::__1::unordered_map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, long, std::__1::hash<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::equal_to<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, long> > > >, DB::MergeTreeData const&, std::__1::vector<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, std::__1::allocator<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > > > const&, bool, Poco::Logger*) @ 0x1605fe85 in /usr/bin/clickhouse
16. DB::ReadFromMergeTree::selectRangesToRead(std::__1::vector<std::__1::shared_ptr<DB::IMergeTreeDataPart const>, std::__1::allocator<std::__1::shared_ptr<DB::IMergeTreeDataPart const> > >) const @ 0x1605edd0 in /usr/bin/clickhouse
17. DB::ReadFromMergeTree::getAnalysisResult() const @ 0x16061abd in /usr/bin/clickhouse
18. DB::ReadFromMergeTree::initializePipeline(DB::QueryPipelineBuilder&, DB::BuildQueryPipelineSettings const&) @ 0x1606200c in /usr/bin/clickhouse
19. DB::ISourceStep::updatePipeline(std::__1::vector<std::__1::unique_ptr<DB::QueryPipelineBuilder, std::__1::default_delete<DB::QueryPipelineBuilder> >, std::__1::allocator<std::__1::unique_ptr<DB::QueryPipelineBuilder, std::__1::default_delete<DB::QueryPipelineBuilder> > > >, DB::BuildQueryPipelineSettings const&) @ 0x1602b0e6 in /usr/bin/clickhouse
20. DB::QueryPlan::buildQueryPipeline(DB::QueryPlanOptimizationSettings const&, DB::BuildQueryPipelineSettings const&) @ 0x1604593c in /usr/bin/clickhouse
21. DB::InterpreterSelectWithUnionQuery::execute() @ 0x14ebcb82 in /usr/bin/clickhouse
22. ? @ 0x151ae649 in /usr/bin/clickhouse
23. DB::executeQuery(DB::ReadBuffer&, DB::WriteBuffer&, bool, std::__1::shared_ptr<DB::Context>, std::__1::function<void (std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&)>, std::__1::optional<DB::FormatSettings> const&) @ 0x151b2af0 in /usr/bin/clickhouse
24. DB::HTTPHandler::processQuery(DB::HTTPServerRequest&, DB::HTMLForm&, DB::HTTPServerResponse&, DB::HTTPHandler::Output&, std::__1::optional<DB::CurrentThread::QueryScope>&) @ 0x15c440e9 in /usr/bin/clickhouse
25. DB::HTTPHandler::handleRequest(DB::HTTPServerRequest&, DB::HTTPServerResponse&) @ 0x15c481f4 in /usr/bin/clickhouse
26. DB::HTTPServerConnection::run() @ 0x15cc3c02 in /usr/bin/clickhouse
27. Poco::Net::TCPServerConnection::start() @ 0x18a2eb93 in /usr/bin/clickhouse
28. Poco::Net::TCPServerDispatcher::run() @ 0x18a2feed in /usr/bin/clickhouse
29. Poco::PooledThread::run() @ 0x18bf2849 in /usr/bin/clickhouse
30. Poco::ThreadImpl::runnableEntry(void*) @ 0x18bf0102 in /usr/bin/clickhouse
31. ? @ 0x7fa6280a8609 in ?
 (version 22.8.15.23 (official build))
```

**Comment by tezhm** [comment]
This issue seems to be fixed in 23.2.4.12, any timeline on when the fix will be released to 23.8.x?

**Comment by den-crane** [comment]
fixed in 22.12.1.1752

* https://github.com/ClickHouse/ClickHouse/pull/43311/files

it helps to disable short_circuit_function_evaluation, unfortunately only with profile level + restart

```
cat /etc/clickhouse-server/users.d/short_circuit_function_evaluation.xml
<?xml version="1.0" ?>
<clickhouse>
    <profiles>
        <default>
            <short_circuit_function_evaluation>disable</short_circuit_function_evaluation>
        </default>
    </profiles>
</clickhouse>
```

```
select value From system.settings where name = 'short_circuit_function_evaluation';
┌─value───┐
│ disable │
└─────────┘

create table test (
                  id Int64,
                  a LowCardinality(Nullable(String)),
                  b LowCardinality(Nullable(String))
                  )
                Engine = MergeTree order by id
                  as  select number%100, 'xxxx', 'yyyy'
                 from numbers(10);

ALTER TABLE test UPDATE a = b WHERE id = 1 SETTINGS mutations_sync = 2;

SELECT * FROM test WHERE id <= 2;
┌─id─┬─a────┬─b────┐
│  0 │ xxxx │ yyyy │
│  1 │ yyyy │ yyyy │
│  2 │ xxxx │ yyyy │
└────┴──────┴──────┘
```

**Comment by den-crane** [comment]
Ah, I mean mutations can be fixed only via default profile. Usual queries can fixed using session or settings  https://fiddle.clickhouse.com/3e92f697-55c3-4db3-b180-f47433ca1a16
